### ISSUE TITLE
E501 false positive when using east_asian_unicode string

### ISSUE THREAD

**Issue by MarcoGorelli** [astral-sh/ruff#3902]
If I have a file with:
```python
from pandas import DataFrame

df = DataFrame(
    {"a": ["あああああ", "い", "う", "えええ"], "b": ["あ", "いいい", "う", "ええええええ"]},
    index=["a", "bb", "c", "ddd"],
)
```
then `ruff` reports a `E501` error (whereas flake8 doesn't)
```console
(.venv) marcogorelli@DESKTOP-U8OKFP3:~/tmp$ ruff t.py
t.py:4:89: E501 Line too long (93 > 88 characters)
Found 1 error.
(.venv) marcogorelli@DESKTOP-U8OKFP3:~/tmp$ flake8 t.py
```

Indeed, if I open the file in Python and count the length of each line, I only get 72 for the offending line:
```python
>>> with open('t.py') as fd:
...     content = fd.read()
...
>>> [len(line) for line in content.splitlines()]
[28, 0, 15, 72, 34, 1]
```

**Comment by charliermarsh** [comment]
This is a consequence of migrating from character-count to unicode-width for line-length violations (https://github.com/charliermarsh/ruff/pull/3714), which was done in part for consistency with Black, which IIUC is moving towards splitting long lines based on width (https://github.com/psf/black/pull/3445). Whether this was a correct decision for us, and in the context of line-length violations, is something worth debating, but just wanted to clarify why and how this changed (since, until recently, we did use character count).

**Comment by MarcoGorelli** [comment]
makes sense, thanks for explaining!

**Comment by charliermarsh** [comment]
(Separately, just realizing that it's confusing to say `(93 > 88 characters)` when we're not actually counting _characters_.)

**Comment by kuba-lilz** [comment]
I see this problem when linting python files that contain comments/docstrings with Japanese characters using ruff

For this line that pylint and all other tools correctly recognize to have 85 characters, ruff==0.0.267 reports it as having 122.
```
    # corrected_best_angleとtick_anglesの関係性を再確認する（tick_anglesの両端から外れてしまう場合について注意深く調べる）
```

**Comment by vincent-olivert-riera** [comment]
I'm having the same issue. Will this get fixed any time soon?

**Comment by charliermarsh** [comment]
@vincent-olivert-riera - I believe you need to run with `--preview`, as the change to use character width is part of the 23.3.0 [preview](https://github.com/psf/black/blob/main/CHANGES.md#2330) style.

**Comment by charliermarsh** [comment]
(I'm going to close this as a duplicate of #3902, although there's useful discussion in both issues.)
